### ISSUE TITLE
Bug Fix: Policy Gradient Computation Using Wrong Model Parameters

### ISSUE THREAD

**Issue by satish860** [SakanaAI/self-adaptive-llms#17]
Description:
There appears to be an implementation issue in the policy gradient computation where we may be using the base model's parameters instead of the modified SVF parameters. This needs to be fixed to properly align with the paper's approach.

## Current Behavior
Currently in `step_optimization`, we:
1. Generate outputs using the modified model through VLLM
2. Compute policy gradients without explicitly ensuring we're using the modified parameters
3. This may lead to incorrect gradient computations since we want to learn how to adjust the modified model's parameters

## Expected Behavior
The policy gradient should be computed using the modified model's probability distribution (πθW'), while only using the base model (πθW) for KL divergence reference when that option is enabled.

## Proposed Fix
Here's the corrected implementation for `step_optimization`:

```python
def step_optimization(self, model_id, model, tokenizer, policy, task_loader, batch_ix, 
                     train_data, train_eval, base_params, decomposed_params, 
                     original_model_params, metrics_to_log, vllm_model=None, **kwargs):
    use_kl_loss = self.use_kl_loss
    kl_ref_coeff = self.kl_ref_coeff
    gpu = self.gpu

    # Get prompts for the batch
    prompts = [task_loader.get_prompt(tokenizer, train_data, i, model_id=model_id) 
              for i in batch_ix]
    clipped_batch_size = len(prompts)

    # Step 1: Generate outputs using modified model through VLLM
    learnable_params = policy.get_learnable_params()
    new_params = forward(policy, model, base_params, decomposed_params, learnable_params)
    load_hf_params_to_vllm(new_params, vllm_model.llm)
    res = eval_model(vllm_model, train_eval, batch_ix)
    rewards = self.get_rewards(task_loader=task_loader, res=res)

    # Step 2: Get reference probabilities from original model if using KL loss
    if use_kl_loss:
        with torch.no_grad():
            # Load original parameters for reference
            load_base_params(model=model, base_params=original_model_params)
            ref_log_probs_list = self.compute_ref_logprobs(
                model=model,
                tokenizer=tokenizer,
                prompts=prompts,
                res=res,
            )
            # Restore modified parameters after getting reference probabilities
            model.load_state_dict(new_params)

    # Step 3: Compute policy gradients using modified model
    model.load_state_dict(new_params)  # Ensure we're using modified parameters
    for j, prompt in enumerate(prompts):
        # ... rest of the policy gradient computation ...
```

## Key Changes
1. Added explicit model parameter management to ensure correct parameters are used at each step
2. Clear separation between:
   - Generation using modified model
   - Reference probability computation using original model (for KL loss)
   - Policy gradient computation using modified model
3. Added explicit loading of modified parameters before computing policy gradients

## Why This Matters
This fix is crucial because:
1. The policy gradient should reflect how to improve the modified model's behavior
2. Using the wrong model parameters could lead to incorrect gradient updates
3. This aligns with the paper's mathematical formulation where policy gradients are computed using πθW'

## Testing Suggestions
1. Verify that policy gradients are computed using modified parameters
2. Check that KL divergence (when enabled) properly measures difference between modified and original models
3. Compare training curves before and after the fix

Would you like me to explain any part of the proposed changes in more detail?

/label bug
/label needs-review

**Comment by floatingsun** [comment]
Can you link me to the code?

**Comment by floatingsun** [comment]
Closing because it looks highly like LLM's output
